src/wasm/wasm-binary.cpp

```diff
@@ -330,18 +330,82 @@ void WasmBinaryWriter::writeExports() {
 
 void WasmBinaryWriter::writeDataSegments() {
   if (wasm->memory.segments.size() == 0) return;
-  uint32_t num = 0;
+  Index numConstant = 0,
+        numDynamic = 0;
   for (auto& segment : wasm->memory.segments) {
-    if (segment.data.size() > 0) num++;
+    if (segment.data.size() > 0) {
+      if (segment.offset->is<Const>()) {
+        numConstant++;
+      } else {
+        numDynamic++;
+      }
+    }
+  }
+  // check if we have too many dynamic data segments, which we can do nothing about
+  auto num = numConstant + numDynamic;
+  if (numDynamic + 1 >= MaxDataSegments) {
+    std::cerr << "too many non-constant-offset data segments, wasm VMs may not accept this binary" << std::endl;
+  }
+  // we'll merge constant segments if we must
+  if (numConstant + numDynamic >= MaxDataSegments) {
+    numConstant = MaxDataSegments - numDynamic - 1;
+    num = numConstant + numDynamic;
+    assert(num == MaxDataSegments - 1);
   }
   auto start = startSection(BinaryConsts::Section::Data);
   o << U32LEB(num);
-  for (auto& segment : wasm->memory.segments) {
-    if (segment.data.size() == 0) continue;
+  // first, emit all non-constant-offset segments; then emit the constants,
+  // which we may merge if forced to
+  Index emitted = 0;
+  auto emit = [&](Memory::Segment& segment) {
     o << U32LEB(0); // Linear memory 0 in the MVP
     writeExpression(segment.offset);
     o << int8_t(BinaryConsts::End);
     writeInlineBuffer(&segment.data[0], segment.data.size());
+    emitted++;
+  };
+  auto& segments = wasm->memory.segments;
+  for (auto& segment : segments) {
+    if (segment.data.size() == 0) continue;
+    if (segment.offset->is<Const>()) continue;
+    emit(segment);
+  }
+  for (Index i = 0; i < segments.size(); i++) {
+    auto& segment = segments[i];
+    if (segment.data.size() == 0) continue;
+    if (!segment.offset->is<Const>()) continue;
+    if (emitted + 2 < MaxDataSegments) {
+      emit(segment);
+    } else {
+      // we can emit only one more segment! merge everything into one
+      // start the combined segment at the bottom of them all
+      auto start = segment.offset->cast<Const>()->value.getInteger();
+      for (Index j = i + 1; j < segments.size(); j++) {
+        auto& segment = segments[j];
+        if (segment.data.size() == 0) continue;
+        if (!segment.offset->is<Const>()) continue;
+        auto offset = segment.offset->cast<Const>()->value.getInteger();
+        start = std::min(start, offset);
+      }
+      // create the segment and add in all the data
+      Const c;
+      c.value = Literal(int32_t(start));
+      c.type = i32;
+      Memory::Segment combined(&c);
+      for (Index j = i; j < segments.size(); j++) {
+        auto& segment = segments[j];
+        if (segment.data.size() == 0) continue;
```

Comment:
These two conditions are repeated three times here. Maybe extract a lambda (or method), `if (isEmptyOrConst(segment)) continue;`?

Reply:
Good idea, done.

Reply:
Oh, I was thinking merge the conditions to reduce line/continue count, i.e. go from
```c++
if (foo.isA()) continue;
if (foo.isB()) continue;
```
to
```c++
if (foo.isAOrB()) continue;
```
There's one condition where which checks both `size() == 0` and NOT `->is<Const>`, so either leave it standalone or also make `isEmptyOrNotConst`.

This is also fairly minor, so Nit

Reply:
Oh, sorry, I didn't understand you fully before.

I kind of prefer it now - with the nice function names, I think it's clear to read. `isAOrB()` seems kind of excessive, at least to me.

Reply:
Yeah the name `isAOrB` is pretty contrived. I think ideally it would be something like `canEmit` or `isSafeToEmit` or `canMerge` or something, where the name implies why we care about A or B. I'm not sure if A and B are actually related though? So I mentioned it so that if you had an idea for what to unify them as, you could name it well. Failing that, a unifying function is probably overkill yeah.

Reply:
Well yeah, actually I'm convinced now. We do only care about one type of segment at that point (and til the end), so why not have a helper function for it. Added.